docs/linux/wifi_fuzzing.md

```diff
@@ -0,0 +1,142 @@
+# Wifi fuzzing
+
+Syzkaller is equipped with necessary functionality and descriptions to fuzz 802.11 subsystem of Linux.
+Currently it focuses on the following two targets.
+- Configuration and management functionality. It is exposed via the nl80211 interface of the Linux kernel.
+- Incoming wireless frames processing. Syzkaller reaches the corresponding code by injecting 802.11 frames. This functionality is under development at the moment.
+
+This document describes the internals of the approach and keeps track of the progress towards its complete implementation.
+
+## General approach
+
+Syzkaller uses mac80211_hwsim module to emulate wifi devices.
+
+In order to simplify reproducer generation and to have an operable 802.11 network from the very beginning, syzkaller performs the following at executor initialization. If `wifi` feature is enabled (it is enabled by default), then:
+1. Two virtual wifi devices are created and assigned predefined MAC addresses (08:02:11:00:00:00 and 08:02:11:00:00:01).
+2. These devices are put into `IBSS` mode.
+3. These devices form an IBSS network. Network parameters are predefined: BSSID (50:50:50:50:50:50), SSID('\x10\x10\x10\x10\x10\x10'), channel (2412 MHz). After this step, the network is operable and 802.11 frame injection is possible.
+
+These steps happen inside the `initialize_wifi_devices` function of `executor/common_linux.h`.
+
+In order to facilitate 802.11 fuzzing, two pseudo syscalls are also introduced.
+1. `syz_80211_inject_frame(mac_addr, buf, buf_len)` performs 802.11 frame injection. The frame will appear to be received at the specified network interface.
+2. `syz_80211_join_ibss(network_interface, ssid, ssid_len, wait_mode)` puts the corresponding network interface into IBSS mode and joins the specified ad-hoc network. Although an IBSS network is already formed at the very beginning, the fuzzer can quickly take interfaces out of operational state (e.g. by deleting and re-adding it), and it will be very problematic for the fuzzer to restore the operational state. Similar syscalls might be added for other wifi modes as well.
+
+## TODO list
```

Comment:
I think we should replace this TODO-list with a pointer to Github issues. That way these aren't entirely lost. You can create a new tag for them ("wifi" ?) and then we could use this link: https://github.com/google/syzkaller/issues?q=is%3Aissue+is%3Aopen+label%3Aos%3Alinux+label%3Awifi

Reply:
I opened an issue for (3) and found an already existing issue that might be useful for (4). (1) and (2) will be fixed by the next PR, so it doesn't make sense to open issues for them.

`docs/linux/external_fuzzing_usb.md` is a similar document and it has a TODO list, so I think it's fine to keep it. The items are present in issues, so it won't be entirely lost.

Reply:
Maybe ``docs/linux/external_fuzzing_usb.md``'s TODOs should be moved to issues. ;-)

(The main benefit of a real issue is that discussions can happen on it, and it's clear if somebody else decides to pick up the work.)

Reply:
Ok, I removed that section from the doc.